lightningd/invoice.c

```diff
@@ -156,105 +158,153 @@ static bool parse_fallback(struct command *cmd,
+	 * resolved in a different order, and can reach
+	 * this part of the code.
+	 * e.g. one might be created with maxroutes>0 and
+	 * a slightly later one with maxroutes>0, both with
+	 * same label.
+	 * The duplicate-label check for the first is done,
+	 * which passes, then the command is suspended while
+	 * waiting for the gossipd to reply with private
+	 * routes.
+	 * The second command with same label will then
+	 * be started, with the same-label check redone,
+	 * but the same-label check will still pass because
+	 * the first command has not yet created the invoice
+	 * yet.
+	 * So we need to recheck duplicate label and
+	 * duplicate rhash here. */
+	if (wallet_invoice_find_by_label(wallet, &invoice, ic->label_val)) {
+		command_fail_detailed(cmd, INVOICE_LABEL_ALREADY_EXISTS,
+				      NULL,
+				      "Duplicate label '%s'",
+				      ic->label_val->s);
+		return;
+	}
+	if (wallet_invoice_find_by_rhash(wallet, &invoice, &ic->rhash)) {
+		command_fail_detailed(cmd, INVOICE_PREIMAGE_ALREADY_EXISTS,
+				      NULL, "preimage already used");
+		return;
 	}
-	/* All cases should have been handled above. */
-	abort();
+
+	/* Fill in the BOLT11 invoice. */
+	b11 = new_bolt11(ic, ic->msatoshi_val);
+	b11->chain = get_chainparams(cmd->ld);
+	b11->timestamp = time_now().ts.tv_sec;
+	b11->payment_hash = ic->rhash;
+	b11->receiver_id = cmd->ld->id;
+	b11->min_final_cltv_expiry = cmd->ld->config.cltv_final;
+	b11->expiry = ic->expiry;
+	b11->description = tal_steal(b11, ic->desc_val);
+	b11->description_hash = NULL;
+	if (ic->fallback_scripts)
+		b11->fallbacks = tal_steal(b11, ic->fallback_scripts);
+	if (ic->routes)
+		b11->routes = tal_steal(b11, ic->routes);
+
+	/* Encode the BOLT11 invoice. */
+	b11enc = bolt11_encode(cmd, b11, false, &hsm_sign_b11, cmd->ld);
+
+	/* Add invoice to database. */
+	result = wallet_invoice_create(wallet,
+				       &invoice,
+				       take(ic->msatoshi_val),
+				       take(ic->label_val),
+				       ic->expiry,
+				       b11enc,
+				       &ic->r,
+				       &ic->rhash);
+	if (!result) {
+		command_fail(cmd, "Failed to create invoice on database");
+		return;
+	}
+
+	/* Get details, then output them. */
+	wallet_invoice_details(cmd, wallet, invoice, &details);
+
+	response = new_json_result(cmd);
+	json_object_start(response, NULL);
+	json_add_hex(response, "payment_hash",
+		     &details.rhash, sizeof(details.rhash));
+	if (deprecated_apis)
+		json_add_u64(response, "expiry_time", details.expiry_time);
+	json_add_u64(response, "expires_at", details.expiry_time);
+	json_add_string(response, "bolt11", details.bolt11);
+	json_object_end(response);
+
+	command_success(cmd, response);
 }
-/* Create 'r' routes for bolt11. */
-static struct route_info **
-create_bolt11_routes(const tal_t *ctx,
-		     struct lightningd *ld,
-		     unsigned int max_routes)
+
+/* Handle the reply from the gossip_get_private_routes request. */
+static void
+invoice_command_get_private_routes_reply(struct subd *_ UNUSED,
```

Comment:
Rather ugly that now even the simple `invoice` command has to become an async command stashing it's state aside...

Reply:
But if it gets the job done

Reply:
An alternative would be for gossipd to send `channel_update`s for local channels to lightningd, and lightningd would then use that information.  It might even be better for lightningd, since it can also directly check channel capacity.